lib/gpu/lib/src/buffer.dart

```diff
@@ -23,6 +23,74 @@ class BufferView {
       {required this.offsetInBytes, required this.lengthInBytes});
 }
 
+/// A buffer that can be referenced by commands on the GPU.
+mixin Buffer {}
+
+/// [DeviceBuffer] is a region of memory allocated on the device heap
+/// (GPU-resident memory).
+base class DeviceBuffer extends NativeFieldWrapperClass1 with Buffer {
+  bool _valid = false;
+  get isValid {
+    return _valid;
+  }
+
+  /// Creates a new DeviceBuffer.
+  DeviceBuffer._initialize(
+      GpuContext gpuContext, StorageMode storageMode, int sizeInBytes)
+      : storageMode = storageMode,
+        sizeInBytes = sizeInBytes {
+    _valid = _initialize(gpuContext, storageMode.index, sizeInBytes);
+  }
+
+  /// Creates a new host visible DeviceBuffer with data copied from the host.
+  DeviceBuffer._initializeWithHostData(GpuContext gpuContext, ByteData data)
+      : storageMode = StorageMode.hostVisible,
+        sizeInBytes = data.lengthInBytes {
+    _valid = _initializeWithHostData(gpuContext, data);
+  }
+
+  final StorageMode storageMode;
+  final int sizeInBytes;
+
+  /// Wrap with native counterpart.
+  @Native<Bool Function(Handle, Pointer<Void>, Int, Int)>(
+      symbol: 'InternalFlutterGpu_DeviceBuffer_Initialize')
+  external bool _initialize(
+      GpuContext gpuContext, int storageMode, int sizeInBytes);
+
+  /// Wrap with native counterpart.
+  @Native<Bool Function(Handle, Pointer<Void>, Handle)>(
+      symbol: 'InternalFlutterGpu_DeviceBuffer_InitializeWithHostData')
+  external bool _initializeWithHostData(GpuContext gpuContext, ByteData data);
+
+  /// Overwrite a range of bytes in the already created [DeviceBuffer].
+  ///
+  /// This method can only be used if the [DeviceBuffer] was created with
+  /// [StorageMode.hostVisible]. An exception will be thrown otherwise.
+  ///
+  /// The entire length of [sourceBytes] will be copied into the [DeviceBuffer],
+  /// starting at byte index [destinationOffsetInBytes] in the [DeviceBuffer].
+  /// If performing this copy would result in an out of bounds write to the
+  /// buffer, then the write will not be attempted and will fail.
+  ///
+  /// Returns [true] if the write was successful, or [false] if the write
+  /// failed due to an internal error.
+  bool overwrite(ByteData sourceBytes, {int destinationOffsetInBytes = 0}) {
```

Comment:
My inclination would be to leave this out, even if usefull, provided we can accomplish the same functionality by recreating device buffers. Synchronization is something we're going to need to think about quite a bit, as well as performance - but it will be easier to do that with less API.

Reply:
This sounds like making DeviceBuffers immutable? I totally get what you're saying here -- we don't _need_ this for getting things drawing. But I'm confident that being able to reuse device allocations is going to be important and is something we should just contend with making work immediately. The port of Scene will use this call.

> Synchronization is something we're going to need to think about quite a bit, as well as performance - but it will be easier to do that with less API.

We talked about this a bit offline, but synchronization shouldn't be an issue. For MVP, all Flutter GPU submissions will happen on the UI thread, and drawing to a canvas would need to happen with a swap chain. I don't think we should mess with trying to support deferring anything until much later on.